Implement `torch._foreach_lerp`

crcrpar opened the pull request:
As per title.

- [ ] ~~Q: Do we want `torch._foreach_lerp.ScalarList` as well?~~
- [ ] ~~we might want to have `ATen/native/cuda/lerp.cuh` and include it in `ATen/native/cuda/Lerp.cu` and `ATen/native/cuda/ForeachTernaryOp.cu`~~

Related:
- https://github.com/pytorch/pytorch/issues/58833
- https://github.com/pytorch/pytorch/issues/71683

cc @vadimkantorov @ptrblck

crcrpar committed:
Implement foreach_lerp

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
add test for foreach_lerp

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

vadimkantorov commented:
Also, unfortuately, it doesn't solve all needed foreach-needs of adam :(
e.g.:
torch._foreach_mul_(acc_deltas, rho)
torch._foreach_addcmul_(acc_deltas, deltas, deltas, value=1 - rho)

But foreach_lerp is useful for a more idiomatic EMA optimizer anyway.

vadimkantorov commented on aten/src/ATen/native/cuda/ForeachTernaryOp.cu:
```diff
@@ -0,0 +1,249 @@
+
+// Ref: [GPU Pro Tip: Lerp Faster in C++](https://developer.nvidia.com/blog/lerp-faster-cuda/)
+template<>
+inline C10_HOST_DEVICE float lerp_(const float start, const float end, const float weight) {
```
Why does it have underscore suffix? As far as I can see, it's not "inplace" in any sense

crcrpar replied:
mimicking PowKernel.cu https://github.com/pytorch/pytorch/blob/233305a852e1cd7f319b15b5137074c9eac455f6/aten/src/ATen/native/cuda/PowKernel.cu#L23-L27

ngimel commented on aten/src/ATen/native/cuda/ForeachTernaryOp.cu:
```diff
@@ -0,0 +1,249 @@
+// Ref: [GPU Pro Tip: Lerp Faster in C++](https://developer.nvidia.com/blog/lerp-faster-cuda/)
+template<>
+inline C10_HOST_DEVICE float lerp_(const float start, const float end, const float weight) {
+  return fma(weight, end, fma(-weight, start, start));
```
this is different than generic implementation above, and than eager implementation, do we want this divergence? Each of the branches in the generic implementation can be rewritten with 2 fmas like here, so maybe we should just do that? For fixed weight there would be no branch divergence

crcrpar replied:
I wanted to make sure the precision of this fma based implementation first by comparing foreach_lerp against at::native::lerp before introducing integrating to the at::native::lerp.
I think I've overlooked some tests which confirms that parity of at::natie::lerp_cpu and at::native::lerp_cuda

crcrpar replied:
so, my question is, would you mind if I update ATen/native/cuda/Lerp.cu to use this fma based logic for certain dtyps?

ngimel replied:
I don't mind fma logic, however, we should keep weight > 0.5 comparison and have 2 fma branches.

ngimel commented on test/test_foreach.py:
```diff
@@ -14,7 +14,7 @@
 from torch.testing._internal.common_methods_invocations import (
     foreach_unary_op_db, foreach_binary_op_db, foreach_pointwise_op_db, foreach_minmax_op_db,
-    foreach_reduce_op_db)
+    foreach_reduce_op_db, foreach_lerp_op_db)
```
why do we need separate foreach_lerp_op_db, can it be part of foreach_pointwise_op_db (which, iirc, has 3 arguments).

crcrpar replied:
the number of arguments and their pattern can be different. pointwise, addcmul and addcdiv, take three lists of tensors and an optional scalar or an optional list of scalars while the arguments of lerp can be two lists of tensors and one list of tensors and one scalar.

ngimel replied:
Regular ops covered by OpDB have very different arguments and yet are all tested by common tests, so I think it should be possible not to multiply dbs here

crcrpar replied:
After checking how the other `OpInfo` tests e.g. test_ops are doing, it seems like I should rely on `SampleInput` https://github.com/pytorch/pytorch/blob/b3206268ace6ebcb5d716ed6673876e62ef484f2/torch/testing/_internal/opinfo/core.py#L114 to better handles different arguments

crcrpar committed:
fix `import torch`

and hopefully build as well.
I have no idea why previous commits did work even without
`<ATen/ops/_foreach_lerp_native.h>`.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Use `SampleInput` for `test_foreach.py`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
tweak n_expected_cudaLaunchKernels

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
remove self-documenting expressions (f-string)

which is a Python3.8 feature...
https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
fix typo

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
This has now integrated `SampleInput` into `ForeachOpInfo` and is ready for another review

vadimkantorov commented:
Another usecase for foreach_lerp is to implement explicit, manual running stats updates for a bunch of batchnorm modules (context in https://github.com/pytorch/pytorch/issues/90342#issuecomment-1341281570).

In this way, the fields running_mean/var of batchnorm modules are used to store current batch mean/var, and then the update of running_mean/var parameters is done by a separate optimizer / manually using foreach_lerp

crcrpar committed:
TEST_WITH_SLOW is used

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
clean up

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
@ngimel friendly ping

crcrpar commented:
@pytorchbot merge

vadimkantorov commented:
@crcrpar Could you please then comment in https://github.com/pytorch/pytorch/issues/71683 what is implemented? Are the remaining things there the EMA optimizer? Btw now with this foreach_lerp, one could do elegant manual updates of BatchNorm stats params (with a manual call to foreach_lerp or to a separate EMA optimizer) and guard against NaN/Inf as one wishes

vadimkantorov commented:
There may be another frequent idiom in optimizers:
https://github.com/pytorch/pytorch/issues/71683#issue-1111826909

that would be fixed by fused op for alpha*tensor1*tensor2 + beta*tensor3*tensor4 (for symmetricitiy): as proposed in https://github.com/pytorch/pytorch/issues/79352#issuecomment-1152946076

crcrpar committed:
update foreach_clamp_(min|max) tests

to use SampleInput

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
@pytorchbot merge

Skylion007 commented on aten/src/ATen/native/ForeachOpsKernels.cpp:
```diff
@@ -283,6 +285,25 @@ FOREACH_POINTWISE_OP_SCALARLIST(addcmul);
+#define FOREACH_TERNARY_OP(OP)                                                                                           \
+std::vector<Tensor> foreach_tensor_ternary_##OP##_slow(TensorList tensors1, TensorList tensors2, TensorList tensors3) {  \
+  check_foreach_api_restrictions(tensors1, tensors2, tensors3);                                                          \
+  std::vector<Tensor> result;                                                                                            \
```
@crcrpar Probably should have done and `result.reserve(tensors1.size());` here.
